sql/tests: TestRandomSyntaxGeneration failed [replication options don't roundtrip parse+format]

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/12148799?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/12148799?buildTab=artifacts#/) on release-23.2 @ [fe7abe63ff6ac18b59971c5c6e21f43f4e8b5f5a](https://github.com/cockroachdb/cockroach/commits/fe7abe63ff6ac18b59971c5c6e21f43f4e8b5f5a):


```
    rsg_test.go:868: 2m10s of 5m0s: 271335 executions, 13943 successful
    rsg_test.go:868: 2m15s of 5m0s: 281561 executions, 14439 successful
    rsg_test.go:868: 2m20s of 5m0s: 292758 executions, 14997 successful
    rsg_test.go:868: 2m25s of 5m0s: 303849 executions, 15522 successful
    rsg_test.go:868: 2m30s of 5m0s: 312336 executions, 15921 successful
    rsg_test.go:868: 2m35s of 5m0s: 320072 executions, 16286 successful
    rsg_test.go:868: 2m40s of 5m0s: 330414 executions, 16818 successful
    rsg_test.go:868: 2m45s of 5m0s: 341534 executions, 17306 successful
    rsg_test.go:868: 2m50s of 5m0s: 353283 executions, 17842 successful
    rsg_test.go:868: 2m55s of 5m0s: 364943 executions, 18340 successful
    rsg_test.go:868: 3m0s of 5m0s: 376471 executions, 18853 successful
    rsg_test.go:868: 3m5s of 5m0s: 386935 executions, 19324 successful
    rsg_test.go:868: 3m10s of 5m0s: 397755 executions, 19861 successful
    rsg_test.go:868: 3m15s of 5m0s: 408888 executions, 20356 successful
    rsg_test.go:868: 3m20s of 5m0s: 420150 executions, 20892 successful
    rsg_test.go:868: 3m25s of 5m0s: 431066 executions, 21409 successful
    rsg_test.go:868: 3m30s of 5m0s: 441701 executions, 21856 successful
    rsg_test.go:868: 3m35s of 5m0s: 452944 executions, 22372 successful
    rsg_test.go:868: 3m40s of 5m0s: 461273 executions, 22728 successful
    rsg_test.go:868: 3m45s of 5m0s: 465574 executions, 22921 successful
    rsg_test.go:868: 3m50s of 5m0s: 470074 executions, 23110 successful
    rsg_test.go:868: 3m55s of 5m0s: 474423 executions, 23275 successful
    rsg_test.go:868: 4m0s of 5m0s: 478793 executions, 23460 successful
    rsg_test.go:868: 4m5s of 5m0s: 483200 executions, 23669 successful
    rsg_test.go:868: 4m10s of 5m0s: 487925 executions, 23892 successful
    rsg_test.go:868: 4m15s of 5m0s: 492357 executions, 24087 successful
    rsg_test.go:868: 4m20s of 5m0s: 496870 executions, 24258 successful
    rsg_test.go:868: 4m25s of 5m0s: 501593 executions, 24467 successful
    rsg_test.go:868: 4m30s of 5m0s: 506347 executions, 24680 successful
    rsg_test.go:868: 4m35s of 5m0s: 510919 executions, 24887 successful
    rsg_test.go:868: 4m40s of 5m0s: 515208 executions, 25081 successful
    rsg_test.go:868: 4m45s of 5m0s: 519494 executions, 25271 successful
    rsg_test.go:868: 4m50s of 5m0s: 523831 executions, 25455 successful
    rsg_test.go:868: 4m55s of 5m0s: 528019 executions, 25632 successful
    rsg_test.go:211: Context cancelled while executing: "REASSIGN OWNED BY LINESTRINGZ , ident TO SIMPLE"
    rsg_test.go:211: Context cancelled while executing: "MOVE ABSOLUTE -259 OVERLAY"
    rsg_test.go:211: Context cancelled while executing: "ANALYZE INDEX_BEFORE_NAME_THEN_PAREN . INNER . REAL"
    rsg_test.go:211: Context cancelled while executing: "CREATE DATABASE IF NOT EXISTS ident"
    rsg_test.go:211: Context cancelled while executing: "COMMENT ON COLUMN FAMILY . WHERE . EXTRACT IS NULL"
    rsg_test.go:211: Context cancelled while executing: "COMMENT ON CONSTRAINT ISERROR ON OVERLAY IS 'string'"
    rsg_test.go:211: Context cancelled while executing: "CREATE DATABASE IF NOT EXISTS ident"
    rsg_test.go:211: Context cancelled while executing: "CREATE DATABASE IF NOT EXISTS ident"
    rsg_test.go:211: Context cancelled while executing: "COMMENT ON SCHEMA INTERVAL . VARCHAR IS 'string'"
    rsg_test.go:211: Context cancelled while executing: "MOVE FORWARD + -498 STREAM"
    rsg_test.go:211: Context cancelled while executing: "DECLARE MULTIPOINT NO SCROLL CURSOR FOR ( ( TABLE error ) )"
    rsg_test.go:904: 532110 executions, 25828 successful
    rsg_test.go:912: cannot parse output of Format: sql="ALTER VIRTUAL CLUSTER INTERVAL 'string' YEAR TO MONTH SET REPLICATION RETENTION = 'string' , RESUME TIMESTAMP = NULL", formattedSQL="ALTER VIRTUAL CLUSTER ('string'::INTERVAL YEAR TO MONTH) SET REPLICATION RETENTION = 'string'RESUME TIMESTAMP = NULL": at or near "resume": syntax error
    rsg_test.go:316: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxGeneration304674583
--- FAIL: TestRandomSyntaxGeneration (301.99s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #111021 sql/tests: TestRandomSyntaxGeneration failed [CALL internal error: expected FuncExpr]  [C-test-failure O-robot O-rsg T-sql-queries branch-master release-blocker]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-32299

The following diff can be used to repro this. (It should be added when fixing this to prevent a regression).
```
diff --git a/pkg/sql/parser/testdata/alter_virtual_cluster b/pkg/sql/parser/testdata/alter_virtual_cluster
index fdb715814fd..5b2aa25a025 100644
--- a/pkg/sql/parser/testdata/alter_virtual_cluster
+++ b/pkg/sql/parser/testdata/alter_virtual_cluster
@@ -408,3 +408,7 @@ ALTER VIRTUAL CLUSTER 'foo' STOP SERVICE -- normalized!
 ALTER VIRTUAL CLUSTER ('foo') STOP SERVICE -- fully parenthesized
 ALTER VIRTUAL CLUSTER '_' STOP SERVICE -- literals removed
 ALTER VIRTUAL CLUSTER 'foo' STOP SERVICE -- identifiers removed
+
+parse
+ALTER VIRTUAL CLUSTER 'foo' SET REPLICATION RETENTION = 'bar' , RESUME TIMESTAMP = NULL
+----
```

repro with `./dev test pkg/sql/parser -f=TestParseDatadriven --rewrite`

The error is 
```
--- FAIL: TestParseDatadriven (0.55s)
    --- FAIL: TestParseDatadriven/alter_virtual_cluster (0.00s)
        parse_test.go:56: at or near "resume": syntax error: ALTER VIRTUAL CLUSTER 'foo' SET REPLICATION RETENTION = 'bar'RESUME TIMESTAMP = NULL
FAIL
```

The problem seems to be that no comma is added when formatting the replication options here: https://github.com/cockroachdb/cockroach/blob/2e0f18b45ff13ddd94f6ac613c5c0f39ed2a9729/pkg/sql/sem/tree/create.go#L2276-L2299

--- a/pkg/sql/parser/testdata/create_virtual_cluster
+++ b/pkg/sql/parser/testdata/create_virtual_cluster
@@ -103,6 +103,14 @@ CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" 
 CREATE VIRTUAL CLUSTER _ FROM REPLICATION OF _ ON 'pgurl' WITH RESUME TIMESTAMP = '132412341234.000000' -- identifiers removed
 
 parse
+CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" ON 'pgurl' WITH RESUME TIMESTAMP = '132412341234.000000', RETENTION = '36h'
+----
+CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" ON 'pgurl' WITH RETENTION = '36h', RESUME TIMESTAMP = '132412341234.000000' -- normalized!
+CREATE VIRTUAL CLUSTER ("destination-hyphen") FROM REPLICATION OF ("source-hyphen") ON ('pgurl') WITH RETENTION = ('36h'), RESUME TIMESTAMP = ('132412341234.000000') -- fully parenthesized
+CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" ON '_' WITH RETENTION = '_', RESUME TIMESTAMP = '_' -- literals removed
+CREATE VIRTUAL CLUSTER _ FROM REPLICATION OF _ ON 'pgurl' WITH RETENTION = '36h', RESUME TIMESTAMP = '132412341234.000000' -- identifiers removed
+
+parse
 CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" ON 'pgurl' WITH OPTIONS (RETENTION = '36h')
 ----
 CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" ON 'pgurl' WITH RETENTION = '36h' -- normalized!

--- a/pkg/sql/sem/tree/create.go
+++ b/pkg/sql/sem/tree/create.go
@@ -2274,7 +2274,15 @@ func (node *CreateTenantFromReplication) Format(ctx *FmtCtx) {
 
 // Format implements the NodeFormatter interface
 func (o *TenantReplicationOptions) Format(ctx *FmtCtx) {
+	var addSep bool
+	maybeAddSep := func() {
+		if addSep {
+			ctx.WriteString(", ")
+		}
+		addSep = true
+	}
 	if o.Retention != nil {
+		maybeAddSep()
 		ctx.WriteString("RETENTION = ")
 		_, canOmitParentheses := o.Retention.(alreadyDelimitedAsSyntacticDExpr)
 		if !canOmitParentheses {
@@ -2286,6 +2294,7 @@ func (o *TenantReplicationOptions) Format(ctx *FmtCtx) {
 		}
 	}
 	if o.ResumeTimestamp != nil {
+		maybeAddSep()
 		ctx.WriteString("RESUME TIMESTAMP = ")
 		_, canOmitParentheses := o.ResumeTimestamp.(alreadyDelimitedAsSyntacticDExpr)
 		if !canOmitParentheses {

